File: client/macos/Votport/LinkPreviewer.swift
import Foundation
import VotportCore

/// Previews the link a field holds: debounced, off the main thread, and a
/// result is kept only while the field still holds the link it was for.
/// The core decides everything about the link; this only carries it.
@MainActor
final class LinkPreviewer: ObservableObject {
    @Published private(set) var preview: LinkPreview?
    @Published private(set) var checking = false
    private var pending: DispatchWorkItem?
    private var current = ""
    /// The kind of link the screen takes; the other kind is named as such.
    private let expect: LinkKind

    init(expect: LinkKind) {
        self.expect = expect
    }

    /// Typing pauses this long before the core is asked.
    private static let settle: TimeInterval = 0.4

    func update(_ link: String) {
        let trimmed = link.trimmingCharacters(in: .whitespacesAndNewlines)
        pending?.cancel()
        pending = nil
        current = trimmed
        guard !trimmed.isEmpty else {
            preview = nil
            checking = false
            return
        }
        checking = true
        let expect = self.expect
        let work = DispatchWorkItem { [weak self] in
            let thread = Thread {
                let result = VotportCore.inspect(link: trimmed, expect: expect)
                DispatchQueue.main.async {
                    MainActor.assumeIsolated { self?.deliver(result, for: trimmed) }
                }
            }
            thread.name = "votport preview"
            thread.start()
        }
        pending = work
        DispatchQueue.main.asyncAfter(deadline: .now() + Self.settle, execute: work)
    }

    private func deliver(_ result: LinkPreview, for link: String) {
        guard link == current else { return }
        preview = result
        checking = false
    }

    /// Whether the primary action may run: the link previewed and is usable.
    var ready: Bool { preview?.usable == true }
    var needsPassword: Bool { preview?.needsPassword == true }
}

/// The one line under a link field, from the core's preview.
@MainActor
enum PreviewLine {
    static func text(_ previewer: LinkPreviewer) -> String? {
        if previewer.checking { return "Checking the link" }
        return previewer.preview?.line
    }

    static func isProblem(_ previewer: LinkPreviewer) -> Bool {
        previewer.preview?.problem != nil
    }
}

File: client/windows/Votport/LinkPreviewer.cs
using Microsoft.UI.Dispatching;
using uniffi.votport_client_core;

namespace Votport;

/// Previews the link a box holds: debounced, off the UI thread, and a result
/// is kept only while the box still holds the link it was for. The core
/// decides everything about the link; this only carries it.
public sealed class LinkPreviewer
{
    private readonly DispatcherQueue dispatcher = DispatcherQueue.GetForCurrentThread();
    private readonly DispatcherQueueTimer timer;
    private readonly Action changed;
    private string current = "";

    internal LinkPreview? Preview { get; private set; }
    public bool Checking { get; private set; }
    public bool Ready => Preview?.Usable == true;
    public bool NeedsPassword => Preview?.NeedsPassword == true;

    /// `changed` runs on the UI thread whenever the preview or its checking
    /// state changes.
    public LinkPreviewer(Action changed)
    {
        this.changed = changed;
        timer = dispatcher.CreateTimer();
        // Typing pauses this long before the core is asked.
        timer.Interval = TimeSpan.FromMilliseconds(400);
        timer.IsRepeating = false;
        timer.Tick += (_, _) => Ask();
    }

    public void Update(string link)
    {
        current = link.Trim();
        timer.Stop();
        if (current.Length == 0)
        {
            Preview = null;
            Checking = false;
            changed();
            return;
        }
        Checking = true;
        changed();
        timer.Start();
    }

    private void Ask()
    {
        var link = current;
        var thread = new Thread(() =>
        {
            LinkPreview? result = null;
            try { result = VotportClientCoreMethods.Inspect(link); }
            catch (Exception e)
            {
                // inspect never fails by contract; a binding mismatch would.
                // The line clears rather than pinning "Checking" forever.
                CrashLog.Append(e);
            }
            dispatcher.TryEnqueue(() =>
            {
                if (link != current) return;
                Preview = result;
                Checking = false;
                changed();
            });
        }) { IsBackground = true, Name = "votport preview" };
        thread.Start();
    }

    /// The one line under a link box, from the core's preview.
    public string? Line()
    {
        if (Checking) return "Checking the link";
        if (Preview is not LinkPreview preview) return null;
        if (preview.Problem is string problem) return problem;
        var parts = new List<string>();
        if (!string.IsNullOrEmpty(preview.Label)) parts.Add(preview.Label);
        if (preview.Kind == LinkKind.Request && preview.MaxBytes is ulong max) parts.Add($"accepts up to {Format.Bytes(max)}");
        if (preview.Kind == LinkKind.Delivery && preview.TotalBytes is ulong total)
        {
            var count = preview.Files.Length;
            parts.Add($"{count} file{(count == 1 ? "" : "s")}, {Format.Bytes(total)}");
        }
        if (preview.NeedsPassword) parts.Add("password needed");
        if (preview.Quic == true) parts.Add("QUIC offered");
        return parts.Count == 0 ? null : string.Join(", ", parts);
    }

    public bool IsProblem => Preview?.Problem is not null;
}
